lib/rules/newline-before-return.js

```diff
@@ -101,10 +101,18 @@ module.exports = function(context) {
      * @private
      */
     function hasNewlineBefore(node) {
-        var tokenBefore = sourceCode.getTokenBefore(node);
-        var lineNumTokenBefore = tokenBefore.loc.end.line;
-        var lineNumNode = node.loc.start.line;
-        var commentLines = calcCommentLines(node, tokenBefore);
+        var tokenBefore = sourceCode.getTokenBefore(node),
+            lineNumNode = node.loc.start.line,
+            lineNumTokenBefore,
+            commentLines;
+
+        if (tokenBefore) {
+            lineNumTokenBefore = tokenBefore.loc.end.line;
+        } else {
+            lineNumTokenBefore = 0;     // assume "just before" start of file
```

Comment:
@kaicataldo To be honest, I wrote that comment because it helped me visualize what we were trying to go for here (I say this knowing that you and I seem to understand the rule a bit differently).

For me, it helped to understand the rules for global return by imagining a "line 0" with a block opening on it, and then we start counting intervening lines after line 0:

```js
// imaginary line 0: {
/* file comment header */
return;
// imaginary closing block I guess
```

But I admit that comment might not be useful to others.